src/mlpack/methods/decision_tree/best_binary_numeric_split.hpp

```diff
@@ -63,13 +60,46 @@ class BestBinaryNumericSplit
       const WeightVecType& weights,
       const size_t minimumLeafSize,
       const double minimumGainSplit,
-      arma::vec& classProbabilities,
+      double& splitInfo,
+      AuxiliarySplitInfo& aux);
+
+  /**
+   * Check if we can split a node.  If we can split a node in a way that
+   * improves on 'bestGain', then we return the improved gain.  Otherwise we
+   * return the value 'bestGain'.  If a split is made, then splitInfo and aux
+   * may be modified.
+   *
+   * It is used only for regression tasks.
+   *
+   * @param bestGain Best gain seen so far (we'll only split if we find gain
+   *      better than this).
+   * @param data The dimension of data points to check for a split in.
+   * @param labels Labels for each point.
+   * @param numClasses Number of classes in the dataset.
+   * @param weights Weights associated with labels.
+   * @param minimumLeafSize Minimum number of points in a leaf node for
+   *      splitting.
+   * @param minimumGainSplit Minimum gain split.
+   * @param splitInfo Stores split information on a successful split.
+   * @param aux Auxiliary split information, which may be modified on a
+   *      successful split.
+   */
+  template<bool UseWeights, typename VecType, typename WeightVecType>
+  static double SplitIfBetter(
+      const double bestGain,
+      const VecType& data,
+      const arma::Row<double>& labels,
+      const size_t numClasses,
```

Comment:
It might be better to make the difference clearer by removing the `numClasses` parameter (since it is irrelevant for regression).  You could also templatize `labels`, since a user could do regression with `float` too.  So I might suggest changing that type to `const ResponsesType& labels` and adding an extra template parameter. :+1:

Reply:
Definitely, I will try to make `ResponseType` a template. 

Actually, the reason I kept numClasses there was just to make its signature the same as the other existing ones. Would it be fine if I change it to `const size_t /* numClasses */`?

Reply:
In this case, I don't think it's necessary to keep the signature the same, since there are really two different overloads here anyway: one for regression and one for classification.

Reply:
I kept the signatures the same because earlier the plan was to write a single `Train` function and use it for both classification and regression.

Reply:
But, if we have to split `DecisionTreeRegressor` into a separate class, then I think we can't have a single `Train` function anyway, right?  So if `DecisionTree` calls a version of `SplitIfBetter()` that does take `numClasses`, and `DecisionTreeRegressor` calls a version of `SplitIfBetter()` that doesn't take a `numClasses` argument, I think that's just fine.

Originally I thought it would be really cool if we could manage to make the `DecisionTree` class do both classification and regression, but as this effort has progressed, I think we came to the conclusion that that's not feasible.

(Please keep in mind you are putting way more time into this than I am, and I'm jumping between lots of different things, so if I am saying something now that contradicts with what I said earlier, I'm really sorry about that!  If I did that, please point it out, because I probably forgot what I said earlier. :cold_sweat:)

Reply:
Yeah, it's fine for me too. I guess I might have misunderstood what you said earlier. I will remove the `numClasses` from the Regressor class. 👍

Reply:
@rcurtin I have applied this change 👍